### ISSUE TITLE
Training Code

### ISSUE THREAD

**Issue by nav99een** [mkara44/liveportrait_talker#9]
Can you please provide the training code? I want to train it for the MEAD dataset to generate emotional and expressive videos.

**Comment by nav99een** [comment]
Please provide the details of the losses you used while training this model. It would be helpful if you could also explain the training strategy.

**Comment by mkara44** [comment]
Hi,

I am planning to share my training code too, however I need time to refactor it. For training, I only trained MappingNet network, which takes 3DMM coefficients as an input and predicts its lip features. I used L1 Loss for training, my ground truth was LivePortrait's Motion Extraction outputs. I didn't use original losses from SadTalker, because of time and cost problems. Those losses would improve the generation accuracy. More information can be found in Sadtalker paper.

Also I don't think that this approach can generate emotional and expressive videos. Because, I used pretrained SadTalker's audio2coeff model, which is forced to focus on only lip coefficients, not facial, during training. You have to train this model too, if you want to generate emotional faces.

**Comment by mkara44** [comment]
Hi again, I added my training pipeline, I will update the ReadMe file too in few days.

**Comment by nav99een** [comment]
Hi, Thanks for sharing the training pipeline. You are using all 64 exp coefficients to predict canonical points only for the lip region. Did you try predicting all 21x3 (expression) canonical points? I was wondering if you had trained the mapping network separately or in an end-to-end manner with the frozen LivePortrait and Sadtalker model because it would have allowed for generating the emotional expression.

**Comment by mkara44** [comment]
Hi, 

MappingNet network is trained to predict only lip related canonical points (6x3). I also tried to predict expression canonical points in the beginning, but it generated inconsistent eyes. This could be result of small dataset, but also I have experienced that Sadtalker may generate inconsistent pupil movements too. Therefore, I decided to use LivePortrait features for eyeblink and pupil movements. 

For your second question, I only trained MappingNet network, other networks were frozen. If I trained audio2exp network too, yes, it would be able to generate emotional expressions too.

**Comment by nav99een** [comment]
Okay, goy it. Thanks for sharing your insights.

**Comment by nav99een** [comment]
Hi again. I am training the mapping_net with your pipeline. I wanted to know how long you trained the model like validation loss or number of epochs.

**Comment by mkara44** [comment]
Hi, I used Early Stopping which stops training if validation loss does not decrease for 10 epochs. Therefore, I did not set the epoch number.
